File: Exercise/Hard/11. Neighbor Zone Maximizer/Code/Neighbor Zone Maximizer.py
def max_zone(matrix):
    if not matrix or not matrix[0]:
        return 0
    
    rows = len(matrix)
    cols = len(matrix[0])
    visited = [[False] * cols for _ in range(rows)]
    max_count = 0

    def is_valid(x, y, target):
        return 0 <= x < rows and 0 <= y < cols and not visited[x][y] and matrix[x][y] == target

    def dfs(x, y, target):
        stack = [(x, y)]
        count = 0
        while stack:
            cx, cy = stack.pop()
            if visited[cx][cy]:
                continue
            visited[cx][cy] = True
            count += 1
            for nx, ny in [(cx-1, cy), (cx+1, cy), (cx, cy-1), (cx, cy+1)]:
                if is_valid(nx, ny, target):
                    stack.append((nx, ny))
        return count

    for i in range(rows):
        for j in range(cols):
            if not visited[i][j]:
                zone_size = dfs(i, j, matrix[i][j])
                if zone_size > max_count:
                    max_count = zone_size

    return max_count

def main():
    matrix = []
    while True:
        line = input().strip()
        if not line:
            break
        row = line.split()
        matrix.append(row)

    result = max_zone(matrix)
    print(result)

if __name__ == "__main__":
    main()

File: Exercise/Hard/11. Neighbor Zone Maximizer/Code/Neighbor Zone Maximizer.cs
using System;
using System.Collections.Generic;

class Program {
    static int MaxZone(List<List<string>> matrix) {
        int rows = matrix.Count;
        if (rows == 0) return 0;
        int cols = matrix[0].Count;
        if (cols == 0) return 0;

        bool[,] visited = new bool[rows, cols];
        int maxCount = 0;

        Func<int, int, string, bool> isValid = (x, y, target) =>
            x >= 0 && x < rows && y >= 0 && y < cols && !visited[x, y] && matrix[x][y] == target;

        Func<int, int, string, int> dfs = null;
        dfs = (x, y, target) => {
            Stack<Tuple<int, int>> stack = new Stack<Tuple<int, int>>();
            stack.Push(new Tuple<int, int>(x, y));
            int count = 0;
            while (stack.Count > 0) {
                var (cx, cy) = stack.Pop();
                if (visited[cx, cy]) continue;
                visited[cx, cy] = true;
                count++;
                int[] dx = { -1, 1, 0, 0 };
                int[] dy = { 0, 0, -1, 1 };
                for (int k = 0; k < 4; ++k) {
                    int nx = cx + dx[k];
                    int ny = cy + dy[k];
                    if (isValid(nx, ny, target)) {
                        stack.Push(new Tuple<int, int>(nx, ny));
                    }
                }
            }
            return count;
        };

        for (int i = 0; i < rows; ++i) {
            for (int j = 0; j < cols; ++j) {
                if (!visited[i, j]) {
                    int zoneSize = dfs(i, j, matrix[i][j]);
                    if (zoneSize > maxCount) {
                        maxCount = zoneSize;
                    }
                }
            }
        }

        return maxCount;
    }

    static void Main() {
 
        List<List<string>> matrix = new List<List<string>>();
        string line;
        while ((line = Console.ReadLine()) != null) {
            if (line.Trim() == "") break;
            string[] parts = line.Trim().Split();
            matrix.Add(new List<string>(parts));
        }
    
        int result = MaxZone(matrix);
        Console.WriteLine(result);
    }
}



